File: files_rpi/tmp/home/tty2rpi/tty2rpi-scripts/game_database.py
import os.path
from log import log

system_map = {
    "adam" : "avision",
    "avision" : "avision",
    "adventurevision" : "avision",
    "arcadia2001" : "co2650",
    "atari2600" : "atari2600",
    "atari5200" : "atari5200",
    "atari7800" : "atari7800",
    "atarilynx" : "atarilynx",
    "ballyastrocade" : "astrocade",
    "astrocade" : "astrocade",
    "genesis" : "genesis",
    "megadrive" : "genesis",
    "coleco" : "coleco",
    "colecovision" : "coleco",
    "channelf" : "channelf",
    "mega-cd" : "megacd",
    "megacd" : "megacd",
    "megaduck" : "megaduck",
    "mister_rbfs" : "arcade",
    "neogeomvs/aes" : "neogeo",
    "neogeo" : "neogeo",
    "nes" : "nes",
    "snes" : "snes",
    "pcengineduo" : "tgfx16",
    "tgfx16" : "tgfx16",
    "playstation" : "psx",
    "psx" : "psx",
    "super32x" : "s32x",
    "s32x" : "s32x",
    "sms" : "sms",
    "supergameboy" : "sgb",
    "sgb" : "sgb",
    "acornarchimedes" : "archie",
    "archie" : "archie",
    "amiga" : "minimig",
    "minimig" : "minimig",
    "atariest/ste" : "atarist",
    "atarist" : "atarist",
    "atom" : "acornatom",
    "acornatom" : "acornatom",
    "bbcmicro/master" : "bbcmicro",
    "bbcmicro" : "bbcmicro",
    "commodore64" : "c64",
    "c64" : "c64",
    "c128" : "c128",
    "electron" : "acornelectron",
    "acornelectron" : "acornelectron",
    "lynx48/96k" : "lynx48",
    "lynx48" : "lynx48",
    "macintoshplus" : "macplus",
    "macplus" : "macplus",
    "pc(486sx)" : "ao486",
    "ao486" : "ao486",
    "pc/xt" : "pcxt",
    "pcxt" : "pcxt",
    "sinclairql" : "ql",
    "ql" : "ql",
    "zxspectrum" : "spectrum",
    "spectrum" : "spectrum",
    "zxspectrumnext" : "zxnext",
    "zxnext" : "zxnext",
    "vectrex" : "vectrex",
    "wonderswan" : "wonderswan",
    "wonderswan color" : "wonderswancolor"
}

def find_system(s):
    words = s.lower().split()
    sys = ''.join(words)
    if sys in system_map:
        return system_map[sys]

    parts = s.lower().split("/")
    for p in parts:
        if p in system_map:
            return system_map[p]

    parts = s.lower().split("_")
    for p in parts:
        if p in system_map:
            return system_map[p]

    for w in words:
        if w in system_map:
            return system_map[w]
    return sys

sys_pics = {}

# rom>title>year>rating>publisher>developer>genre>score>players>description>
ROM = 0
TITLE = 1
YEAR = 2
RATING = 3
PUBLISHER = 4
DEVELOPER = 5
GENRE = 6
SCORE = 7
PLAYERS = 8
DESCRIPTION = 9
NUM_ENTRIES = 10

class GameData:
    _picture = ''

    def __init__(self, system, row):
        self._system  = find_system(system)
        self._entries = row.split('>')
        while len(self._entries) < NUM_ENTRIES:
            self._entries.append('')
        for i in range(0, NUM_ENTRIES - 1):
            self._entries[i] = self._entries[i].strip()

    def _field(self, f):
        return self._entries[f]

    def system(self):
        return self._system

    def set_picture(self, p):
        self._picture = p

    def picture(self):
        return self._picture

    def system_picture(self):
        global sys_pics
        if self.system() in sys_pics:
            return sys_pics[self.system()]
        else:
            return ''

    def log(self):
        log(self._system + ': ' + '>'.join(self._entries))

    def _entry_info(self, s, name, indx):
        if len(self._entries) >= indx and self._field(indx):
            s.append(name + ": ")
            s.append(self._field(indx))

    def info_str(self):
        s = []
        self._entry_info(s, "Title", TITLE)
        self._entry_info(s, "Year", YEAR)
        self._entry_info(s, "Developer", DEVELOPER)
        self._entry_info(s, "Genre", GENRE)
        self._entry_info(s, "Players", PLAYERS)
        self._entry_info(s, "Score", SCORE)

        if len(s) == 2: # No label for just one line
            return s[1].strip()

        str = ''
        for a in range(0, int(len(s) / 2)):
            str = str + s[a * 2] + s[a * 2 + 1] + "\n"

        return str.strip()

class GameDatabase:
    _systems = {}
    _roms = {}
    _titles = {}

    def _scan_pics(self, dir, ext):
        pics = {}
        for fname in os.listdir(dir):
            if fname.endswith("." + ext):
                base = os.path.splitext(fname)[0]
                pics[base] = fname
        return pics

    def __init__(self, dirname, picture_dir, syspic_dir, ext):
        self._pictures = self._scan_pics(picture_dir, ext)
        self._sys_pictures = self._scan_pics(syspic_dir, ext)
        self._scandir(dirname)
        self._active_curpath = ''
        gd = GameData("MENU", "MENU>MENU>MENU")
        gd.set_picture("MENU." + ext)
        self._add(gd)

    def _add(self, gd):
        self._systems[gd.system()] = gd.system()
        self._roms[gd._field(ROM)] = gd
        self._titles[gd._field(TITLE)] = gd
        if gd._field(ROM) in self._pictures:
            gd.set_picture(self._pictures[gd._field(ROM)])

    def _add_file(self, dirname, filename):
        with open(dirname + "/" + filename, 'r') as fp:
            lines = fp.readlines()
            base = os.path.splitext(filename)[0]
            for line in lines:
                gd = GameData(base, line.strip())
                self._add(gd)

    def _scandir(self, dirname):
        for fname in os.listdir(dirname):
            if fname.endswith(".csv"):
                self._add_file(dirname, fname)
        # Always use the MISTER rbfs as the priority
        self._add_file(dirname, "MISTER_rbfs.csv")

    def _decode(self, s):
        return s.replace("%20", " ").replace("%2C", ",").strip()

    def _title_from_startpath(self, sp):
        return os.path.splitext(os.path.basename(os.path.normpath(sp)))[0]

    def lookup_system_pic(self, sys):
        if sys in self._sys_pictures:
            return self._sys_pictures[sys]
        return None

    def _extract_name(self, strs, indx):
        if len(strs) >= indx + 1:
            return self._decode(strs[indx])
        return ''

    def _lookup_title_variants(self, title):
        # Keep trying shortened versions of the title
        tit = title.split()
        title_lookup = None
        for w in range(len(tit) - 1, 0, -1):
            t = ' '.join(tit[0:w])
            log("Trying title '{}'".format(t))
            look = None
            if t in self._titles:
                look = self._titles[t]
            if look:
                if look.picture:
                    title_lookup = look
                    break
                elif not title_lookup:
                    title_lookup = look
        return title_lookup

    def lookup(self, s):
        strs = s.split(",")
        corename  = self._extract_name(strs, 1)
        fullpath  = self._extract_name(strs, 2)
        curpath   = self._extract_name(strs, 3)
        startpath = self._extract_name(strs, 4)
        select    = self._extract_name(strs, 5)
        samgame   = self._extract_name(strs, 6)

        log("corename = {}, fullpath = {}, curpath = {}, startpath = {}, select = {}, samgame = {}".
             format(corename, fullpath, curpath, startpath, select, samgame))

        if curpath.endswith(".mra") or \
           curpath.endswith(".rbf") or \
           curpath.endswith(".nes"):
            curpath = os.path.splitext(curpath)[0]

        core, title, alt_title, system = '', '', '', ''

        if startpath == '/tmp/SAM_game.mgl' and select != 'active':
            # SAM is active : use corename & samgame
            system = "arcade" if corename in self._roms else find_system(corename)
            core = corename
            title = samgame
        elif select == 'active' or select == 'selected':
            # Menu is active, use curpath & fullpath
            core = ''
            title = curpath
            system = "arcade" if "arcade" in fullpath.lower() else find_system(fullpath.lower())
 #       elif select == 'selected':
 #           # Game is active, use corename & startpath and if needed curpath
 #           core = corename
 #           title = self._title_from_startpath(startpath)
 #           alt_title = curpath
 #           self._active_curpath = curpath
 #           if "arcade" in startpath.lower() or "arcade" in fullpath.lower():
 #               system = "arcade"
 #           else:
 #               system = find_system(corename)
        elif select == 'cancelled':
            # Game is active, use corename & startpath and if needed _active_curpath
            system = "arcade" if corename in self._roms else find_system(corename)
            core = corename
            title = self._title_from_startpath(startpath)
            alt_title = self._active_curpath

        log("System = '{}', Core = '{}', Title = '{}', AltTitle = '{}'".format(system, core, title, alt_title))

        core_lookup, title_lookup, alt_title_lookup = None, None, None
        if core and core != "MENU" and core in self._roms:
            core_lookup = self._roms[core]
        if title and title in self._titles:
            title_lookup = self._titles[title]
        if alt_title and alt_title in self._titles:
            alt_title_lookup = self._titles[alt_title]

        if not title_lookup and not alt_title_lookup and (title or alt_title):
            if title:
                title_lookup = self._lookup_title_variants(title)
            if not title_lookup and alt_title:
                alt_title_lookup = self._lookup_title_variants(alt_title)

        if core_lookup:
            log("Core  Lookup = \n{}".format(core_lookup.info_str()))
        if title_lookup:
            log("Title Lookup = \n{}".format(title_lookup.info_str()))
        if alt_title_lookup:
            log("ALT   Lookup = \n{}".format(alt_title_lookup.info_str()))

        if core_lookup:
            return core_lookup, system
        if title_lookup:
            return title_lookup, system
        if alt_title_lookup:
            return alt_title_lookup, system

        # Make a db entry from the lookup key
        log("FALLBACK '{}' '{}' '{}'".format(system, corename, title))
        gd = GameData(system, corename + ">" + title + ">")
        return gd, system

File: files_rpi/tmp/home/tty2rpi/tty2rpi-cpp/GameDatabase.cpp
/*
 * The MIT License (MIT)

 * Copyright (c) 2023 Gary Sweet
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

#include "GameDatabase.h"
#include "StringManip.h"

#include <filesystem>
#include <iostream>

using namespace std;

map<string, string> s_systemAltMap = {
    { "acornarchimedes", "archie" },
    { "acornatom", "acornatom" },
    { "acornelectron", "acornelectron" },
    { "adam", "avision" },
    { "adventurevision", "avision" },
    { "amiga", "minimig" },
    { "ao486", "ao486" },
    { "arcadia2001", "co2650" },
    { "archie", "archie" },
    { "astrocade", "astrocade" },
    { "atari2600", "atari2600" },
    { "atari5200", "atari5200" },
    { "atari7800", "atari7800" },
    { "atariest/ste", "atarist" },
    { "atarilynx", "atarilynx" },
    { "atarist", "atarist" },
    { "atom", "acornatom" },
    { "avision", "avision" },
    { "ballyastrocade", "astrocade" },
    { "bbcmicro", "bbcmicro" },
    { "bbcmicro/master", "bbcmicro" },
    { "c128", "c128" },
    { "c64", "c64" },
    { "channelf", "channelf" },
    { "coleco", "coleco" },
    { "colecovision", "coleco" },
    { "commodore64", "c64" },
    { "electron", "acornelectron" },
    { "genesis", "genesis" },
    { "lynx48", "lynx48" },
    { "lynx48/96k", "lynx48" },
    { "macintoshplus", "macplus" },
    { "macplus", "macplus" },
    { "mega-cd", "megacd" },
    { "megacd", "megacd" },
    { "megadrive", "genesis" },
    { "megaduck", "megaduck" },
    { "minimig", "minimig" },
    { "mister_rbfs", "arcade" },
    { "neogeo", "neogeo" },
    { "neogeomvs/aes", "neogeo" },
    { "nes", "nes" },
    { "pc(486sx)", "ao486" },
    { "pc/xt", "pcxt" },
    { "pcengineduo", "tgfx16" },
    { "pcxt", "pcxt" },
    { "playstation", "psx" },
    { "psx", "psx" },
    { "ql", "ql" },
    { "s32x", "s32x" },
    { "sgb", "sgb" },
    { "sinclairql", "ql" },
    { "sms", "sms" },
    { "snes", "snes" },
    { "spectrum", "spectrum" },
    { "super32x", "s32x" },
    { "supergameboy", "sgb" },
    { "tgfx16", "tgfx16" },
    { "vectrex", "vectrex" },
    { "wonderswan color", "wonderswancolor" },
    { "wonderswan", "wonderswan" },
    { "zxnext", "zxnext" },
    { "zxspectrum", "spectrum" },
    { "zxspectrumnext", "zxnext" },
};

static string FindSystem(const string &str)
{
    string lower = ToLower(str);

    auto Lookup = [&](const string &s){
        auto lookup = s_systemAltMap.find(s);
        if (lookup != s_systemAltMap.end())
            return lookup->second;
        return string();
    };

    auto TokenizedFind = [&](char sep){
        vector<string> parts = Tokenize(lower, sep);

        string ret;
        for (const string &p : parts)
        {
            ret = Lookup(p);
            if (!ret.empty())
                return ret;
        }
        return string();
    };

    string sys, ret;

    vector<string> parts = Tokenize(lower, ' ');
    for (const string &p : parts)
        sys += p;
    ret = Lookup(sys);
    if (!ret.empty())
        return ret;

    ret = TokenizedFind('/');
    if (ret.size() > 0)
        return ret;

    ret = TokenizedFind('_');
    if (ret.size() > 0)
        return ret;

    ret = TokenizedFind(' ');
    if (ret.size() > 0)
        return ret;

    ret = "";
    for (const string &p : parts)
        ret += p;

    return ret;
}

GameRecord::GameRecord(const string &system, const string &dbRow)
{
    m_system = FindSystem(system);

    m_entries = Tokenize(dbRow, '>');
    m_entries.resize(Field::NUM_FIELDS);

    for (auto &e : m_entries)
        Trim(e);
}

string GameRecord::InfoStr() const
{
    string s, t;

    auto add = [&](uint32_t f, const std::string &name) {
        if (!m_entries[f].empty())
        {
            s += name + ": ";
            s += m_entries[f];
            s += "\n";
        }
    };

    if (!m_entries[TITLE].empty())
        t += m_entries[TITLE];

    add(YEAR, "Year");
    add(GENRE, "Genre");
    add(PLAYERS, "Players");
    add(DEVELOPER, "Developer");
    if (m_entries[DEVELOPER] != m_entries[PUBLISHER])
        add(PUBLISHER, "Publisher");
    add(SCORE, "Score");

    if (!s.empty())
        s = "Title: " + t + "\n" + s;
    else
        s = t;

    return Trim(s);
}

template <typename T>
void ScanPicDir(const string &dir, const string &ext, T *container)
{
    for (const auto &file : filesystem::directory_iterator(dir))
    {
        const string &p = file.path();
        if (EndsWith(p, ext))
        {
            vector<string> parts = Tokenize(p.substr(0, p.size() - ext.size()), '/');
            const string &base = ToLower(parts.back());
            (*container)[base] = p;
        }
    }
}

GameRecord *GameDatabase::AddRecord(const string &system, const string &row)
{
    m_games.emplace_back(system, row);

    GameRecord *gr = &m_games.back();

    const string &rom   = ToLower(gr->Field(GameRecord::ROM));
    const string &title = ToLower(gr->Field(GameRecord::TITLE));

    m_romMap[rom]     = gr;
    m_titleMap[title] = gr;

    const auto iter = m_pictures.find(rom);
    if (iter != m_pictures.end())
        gr->SetPicture(iter->second);

    return gr;
}

void GameDatabase::AddFile(const string &path)
{
    const uint32_t BUFSZ = 32 * 1024;

    FILE *fp = fopen(path.c_str(), "r");
    if (!fp)
        throw "Failed to open database file";

    vector<string> parts = Tokenize(path.substr(0, path.size() - 4), '/');
    const string &system = parts.back();

    char buf[BUFSZ];
    while (fgets(buf, BUFSZ, fp))
        AddRecord(system, buf);

    fclose(fp);
}

GameDatabase::GameDatabase(const string &dirname, const string &pictureDir,
                           const string &syspicDir, const string &controllerDir,
                           const string &ext, bool log) :
    m_log(log)
{
    // Scan the pictures
    ScanPicDir(pictureDir, ext, &m_pictures);
    ScanPicDir(syspicDir, ext, &m_sysPictures);
    ScanPicDir(controllerDir, ext, &m_controllerPictures);

    // Scan the database files
    for (const auto &file : filesystem::directory_iterator(dirname))
    {
        const string &p = file.path();
        if (EndsWith(p, ".csv"))
            AddFile(p);
    }
    // Always use the MISTER rbfs as the priority
    AddFile(dirname + "/MISTER_rbfs.csv");

    // Add a dummy menu entry
    GameRecord *rec = AddRecord("MENU", "MENU>MENU");
    rec->SetPicture("MENU" + ext);
}

string GameDatabase::LookupSystemPic(const string &system) const
{
    auto iter = m_sysPictures.find(FindSystem(system));
    if (iter != m_sysPictures.end())
        return iter->second;

    return string();
}

string GameDatabase::LookupControllerPic(const string &system) const
{
    auto iter = m_controllerPictures.find(FindSystem(system));
    if (iter != m_controllerPictures.end())
        return iter->second;

    return string();
}

static string Decode(const string &str)
{
    string s = str;
    Trim(s);

    ReplaceInPlace(s, "%20", " ");
    ReplaceInPlace(s, "%2C", ",");

    return s;
}

static string ExtractName(const vector<string> &strs, uint32_t indx)
{
    if (strs.size() >= indx + 1)
        return Decode(strs[indx]);
    return string();
}

static string TitleFromStartPath(const string &sp)
{
    vector<string> parts = Tokenize(sp, '/');
    string s = parts.back();
    StripExtension(s);
    return s;
}

std::map<std::string, const GameRecord *>::iterator
GameDatabase::LookupVariantsWithPic(const string &title)
{
    // Try shortened versions of the title
    vector<string> parts = Tokenize(ToLower(title), ' ');

    for (size_t p = parts.size() - 1; p > 1; p--)
    {
        string s;
        for (size_t pp = 0; pp < p; pp++)
            s += parts[pp] + " ";
        s[s.size() - 1] = '\0';

        if (m_log)
            cout << "Trying alternate: '" << s << "'\n";

        auto iter = m_titleMap.find(s);
        if (iter != m_titleMap.end())
        {
            const GameRecord &rc = *(iter->second);
            if (!rc.Picture().empty())
                return iter;
        }
    }

    return m_titleMap.end();
}

GameRecord GameDatabase::Lookup(const string &key, string *sys)
{
    vector<string> strs = Tokenize(key, ',');

    string corename  = ExtractName(strs, 1);
    string fullpath  = ExtractName(strs, 2);
    string curpath   = ExtractName(strs, 3);
    string startpath = ExtractName(strs, 4);
    string select    = ExtractName(strs, 5);
    string samgame   = ExtractName(strs, 6);

    StripExtension(curpath);

    string core, title, altTitle, system;
    if (startpath == "/tmp/SAM_game.mgl" && select != "active")
    {
        // SAM is active : use corename & samgame
        if (m_romMap.find(corename) != m_romMap.end())
            system = "arcade";
        else
            system = FindSystem(corename);

        core  = corename;
        title = samgame;
    }
    // "selected" is to ensure the menu doesn't change when the game is loading
    else if (select == "active" || select == "selected")
    {
        string lfp = ToLower(fullpath);
        if (lfp == "_console" || lfp == "_computer")
            lfp = ToLower(curpath);

        // Menu is active, use curpath & fullpath
        if (lfp.find("arcade") != string::npos)
            system = "arcade";
        else
            system = FindSystem(lfp);

        core = "";
        title = curpath;
    }
    else if (select == "cancelled")
    {
        // Game is active, use corename & startpath
        if (m_romMap.find(corename) != m_romMap.end())
            system = "arcade";
        else
            system = FindSystem(corename);

        core = corename;
        title = TitleFromStartPath(startpath);
    }

    if (m_log)
        cout << "Sys = '" << system << "', Core = '" << core <<
                "', Title = '" << title << "', Alt = '" << altTitle << "'\n";

    std::map<std::string, const GameRecord *>::iterator coreLookup     = m_romMap.end();
    std::map<std::string, const GameRecord *>::iterator titleLookup    = m_titleMap.end();
    std::map<std::string, const GameRecord *>::iterator altTitleLookup = m_titleMap.end();

    if (core != "" && core != "menu")
        coreLookup = m_romMap.find(ToLower(core));

    if (title != "")
        titleLookup = m_titleMap.find(ToLower(title));

    if (altTitle != "")
        altTitleLookup = m_titleMap.find(ToLower(altTitle));

    if (titleLookup == m_titleMap.end() && altTitleLookup == m_titleMap.end() &&
        (title != "" || altTitle != ""))
    {
        if (title != "")
            titleLookup = LookupVariantsWithPic(title);
        if (titleLookup == m_titleMap.end() && altTitle != "")
            altTitleLookup = LookupVariantsWithPic(altTitle);
    }

    if (m_log)
    {
        if (coreLookup != m_romMap.end())
            cout << "Core Lookup = \n" << coreLookup->second->InfoStr() << "\n";

        if (titleLookup != m_titleMap.end())
            cout << "Title Lookup = \n" << titleLookup->second->InfoStr() << "\n";

        if (altTitleLookup != m_titleMap.end())
            cout << "Alt Lookup = \n" << altTitleLookup->second->InfoStr() << "\n";
    }

    *sys = system;

    if (coreLookup != m_romMap.end())
        return *coreLookup->second;
    if (titleLookup != m_titleMap.end())
        return *titleLookup->second;
    if (altTitleLookup != m_titleMap.end())
        return *altTitleLookup->second;

    // Make a fallback db entry from the lookup key
    if (m_log)
        cout << "FALLBACK '" << system << "' '" << corename << "' '" << title << "'\n";

    return GameRecord(system, corename + ">" + title + ">");
}
